ext/rapidjson/json_escape.h

```diff
@@ -0,0 +1,90 @@
+#include "ruby.h"
+#include "ruby/encoding.h"
+
+/* strlen("\\u2029") == 6 */
+#define ESCAPE_JSON_MAX_LEN 6
+
+static inline long
+json_escaped_length(VALUE str)
+{
+    const long len = RSTRING_LEN(str);
+    if (len >= LONG_MAX / ESCAPE_JSON_MAX_LEN) {
+        ruby_malloc_size_overflow(len, ESCAPE_JSON_MAX_LEN);
+    }
+    return len * ESCAPE_JSON_MAX_LEN;
+}
+
+static VALUE
+escape_json(VALUE self, VALUE str)
+{
+    if (!RB_TYPE_P(str, T_STRING)) {
+        str = rb_convert_type(str, T_STRING, "String", "to_s");
+    }
+
+    rb_encoding *enc = rb_enc_get(str);
+    if (enc != rb_utf8_encoding() && enc != rb_usascii_encoding()) {
+        rb_raise(rb_eEncCompatError, "input string must be UTF-8 or ASCII");
+    }
+
+    const char *cstr = RSTRING_PTR(str);
+    const unsigned long str_len = RSTRING_LEN(str);
+    const char *end = cstr + RSTRING_LEN(str);
+
+    size_t initial_match = strcspn(cstr, "&<>\xe2");
+    if (initial_match == str_len) {
+        return str;
+    }
+
+    VALUE escaped = rb_str_buf_new(json_escaped_length(str));
```

Comment:
Here you could only do the `* 6` on what's left to match, but not sure if worth it.

Reply:
It's a good idea. I'd really love to be more conservative about the size of buffer allocated, I think if we only reserved a little extra we could avoid the realloc which happens at the end (but in exchange we'd need to detect and possibly perform a realloc in the middle of escaping). I'll think more on it before proposing this somewhere upstream.

Reply:
IIRC `CGI.escape_html` does what I suggest here. However their factor is lower, I think the largest escape is 4bytes. With 6bytes, it's a large increase yeah.

Reply:
It's the same length. I know because I copied it 😆